text/0000-unsafe-fields.md

```diff
@@ -0,0 +1,854 @@
+        // SAFETY: Moving `arc` out of `Self` releases it from its safety
+        // invariant.
+        let arc = unsafe { self.arc };
+        Arc::into_inner(arc).map(UnsafeCell::into_inner)
+    }
+
+    /// Produces a `KeepAlive` handle, which defers the destruction
+    /// of the enclosed value.
+    pub fn keep_alive(&self) -> KeepAlive<T> {
+        // SAFETY: By invariant on `KeepAlive::arc`, this clone will never be
+        // used for accessing `T`, as required by `UniqueArc::arc`. The one
+        // exception is that, if a `KeepAlive` is the last reference to be
+        // dropped, then it will drop the inner `T`. However, if this happens,
+        // it means that the `UniqueArc` has already been dropped, and so its
+        // invariant will not be violated.
+        unsafe {
+            KeepAlive {
+                arc: self.arc.clone(),
+            }
+        }
+    }
+}
+
+impl<T> Deref for UniqueArc<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        // SAFETY: We do not create any other owning references to `arc` - we
+        // only dereference it below, but do not clone it.
+        let arc = unsafe { &self.arc };
+        let ptr = UnsafeCell::get(arc);
+        // SAFETY: We satisfy all requirements for pointer-to-reference
+        // conversions [1]:
+        // - By invariant on `&UnsafeCell<T>`, `ptr` is well-aligned, non-null,
+        //   dereferenceable, and points to a valid `T`.
+        // - By invariant on `Self::arc`, no other `Arc` references exist to
+        //   this value which will be used for reading or writing. Thus, we
+        //   satisfy the aliasing invariant of `&` references.
+        //
+        // [1] https://doc.rust-lang.org/1.85.0/std/ptr/index.html#pointer-to-reference-conversion
+        unsafe { &*ptr }
+    }
+}
+
+impl<T> DerefMut for UniqueArc<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        // SAFETY: We do not create any other owning references to `arc` - we
+        // only dereference it below, but do not clone it.
+        let arc = unsafe { &mut self.arc };
+        let val = UnsafeCell::get(arc);
+        // SAFETY: We satisfy all requirements for pointer-to-reference
+        // conversions [1]:
+        // - By invariant on `&mut UnsafeCell<T>`, `ptr` is well-aligned,
+        //   non-null, dereferenceable, and points to a valid `T`.
+        // - By invariant on `Self::arc`, no other `Arc` references exist to
+        //   this value which will be used for reading or writing. Thus, we
+        //   satisfy the aliasing invariant of `&mut` references.
+        //
+        // [1] https://doc.rust-lang.org/1.85.0/std/ptr/index.html#pointer-to-reference-conversion
+        unsafe { &mut *val }
+    }
+}
+
+// SAFETY: `UniqueArc<T>` has the same aliasing and synchronization properties
+// as `&mut T`, and so it is `Sync` if `&mut T` is `Sync`.
+unsafe impl<T> Sync for UniqueArc<T> where &'static mut T: Sync {}
```

Comment:
this unnecessarily restricts `T: 'static` due to the `&'static mut T`

Reply:
Good point. https://github.com/rust-lang/rfcs/pull/3458/commits/a5cc19f445dd9b18d6d8061cf932d4aa64302cec